### PR TITLE
Singletons

### PR DESCRIPTION
**Pull request by Noviny**
This pull request introduces [Singletons](https://en.wikipedia.org/wiki/Singleton_pattern) as a new feature for configuring global, editable database entries when it's not necessary to have more than one representation of the schema (aka, not a normal list).

Singletons can be useful for things like: 

- System Settings
- Homepage Layout
- Site Navigation

Before this feature, developers could write a typical schema for a list, and then using Keystone hooks limit the number of items within that list to approximate this functionality.
Unfortunately the GraphQL schema didn't reflect these constraints,  and neither did the AdminUI.

This pull requests brings that functionality native to Keystone's schema configuration as a first class primitive.

### Tasks
- [ ] Add enforcement of the databases (where able) to ensure that the singleton list only ever has one item
- [x] Alter the graphql layer so that queries on singletons make more sense (remove querying for multiple items, remove the ‘where’ query on fetching a single item, as there will only ever be one)
- [x] Update the Admin UI - mostly subtractive changes to give a good experience for singletons where navigating items UI is removed, removing clicks

### CODE REVIEW

**Comment by mattJurz** [comment]
I am really looking forward to this feature. Are you able to estimate when singletons will be available in keystone?

File: docs/pages/docs/guides/singletons.mdx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import { RelatedContent } from '../../../components/RelatedContent';
+
+# Singletons
+
+In Keystone, a Singleton is a list type that contains only one entry.
+
+## Configuration
+
+A little about the configuration differences
+
```
Comment (Noviny) [inline docs/pages/docs/guides/singletons.mdx:12]: A thing to call out somewhere, singletons can have relationships *to* things, but can't have relationships *from* something *to* a singleton - this also means that all relationships including a singleton are one-sided only.

File: packages/core/src/lib/createSystem.ts [hunk 2]
```diff
@@ -115,3 +118,16 @@ export function createSystem(config: KeystoneConfig, isLiveReload?: boolean) {
...
+  for (const [, { listKey }] of Object.entries(lists).filter(
+    ([, { kind }]) => kind === 'singleton'
+  )) {
+    const model = prismaClient[listKey[0].toLowerCase() + listKey.slice(1)];
+    const item = await model.findUnique({ where: { id: 1 } });
+
+    if (!item) {
+      await model.create({ data: { id: 1 } });
+    }
+  }
```
Comment (mitchellhamilton) [inline packages/core/src/lib/createSystem.ts:132]: Should use `Promise.all`

File: packages/core/src/schema/schema.ts [hunk 3]
```diff
@@ -19,8 +20,15 @@ export function config<TypeInfo extends BaseKeystoneTypeInfo>(config: KeystoneCo
...
-  return config;
+>(config: ListConfigWithoutKind<ListTypeInfo, Fields>): ListConfig<ListTypeInfo, any> {
+  return { ...config, kind: 'list' };
+}
+
+export function singleton<
+  Fields extends BaseFields<ListTypeInfo>,
+  ListTypeInfo extends BaseSingletonTypeInfo
+>(config: Omit<SingletonConfig<ListTypeInfo, Fields>, 'kind'>): SingletonConfig<ListTypeInfo, any> {
+  return { ...config, kind: 'singleton', db: { ...config.db, idField: { kind: 'autoincrement' } } };
```
Comment (dcousens) [inline packages/core/src/schema/schema.ts:31]: Move this internally to `initialiseLists` (or around that)

File: packages/core/src/lib/core/graphql-schema.ts [hunk 4]
```diff
@@ -28,24 +29,34 @@ export function getGraphQLSchema(
...
         return mutations;
       }),
-      extraFields.mutation
+      extra.mutation
     ),
   });
   const graphQLSchema = new GraphQLSchema({
     query: query.graphQLType,
     mutation: mutation.graphQLType,
     // not about behaviour, only ordering
```
Comment (dcousens) [inline packages/core/src/lib/core/graphql-schema.ts:47]: - [  ] This comment isn't accurate now

File: packages/core/src/lib/core/mutations/access-control.ts [hunk 5]
```diff
@@ -19,7 +19,7 @@ const missingItem = (operation: string, uniqueWhere: UniquePrismaFilter) =>
   );
 
 async function getFilteredItem(
-  list: InitialisedList,
+  list: InitialisedSchema,
```
Comment (dcousens) [inline packages/core/src/lib/core/mutations/access-control.ts:22]:
```suggestion
  schema: InitialisedSchema,
```

File: packages/core/src/lib/core/mutations/index.ts [hunk 6]
```diff
@@ -20,9 +20,37 @@ function promisesButSettledWhenAllSettledAndInOrder<T extends Promise<unknown>[]
...
+  const updateOne = graphql.field({
+    type: list.types.output,
+    args: {
+      where: graphql.arg({ type: graphql.nonNull(list.types.uniqueWhere) }),
+      data: graphql.arg({ type: graphql.nonNull(list.types.update) }),
+    },
+    resolve(_rootVal, args, context) {
+      return createAndUpdate.updateOne(args, list, context);
+    },
+  });
```
Comment (dcousens) [inline packages/core/src/lib/core/mutations/index.ts:52]: noop, move this back

File: packages/core/src/lib/core/queries/resolvers.ts [hunk 7]
```diff
@@ -106,6 +106,18 @@ export async function findOne(
...
 }
 
+export async function findSingleton(list: InitialisedSingleton, context: KeystoneContext) {
+  // Check operation permission to pass into single operation
+  const operationAccess = await getOperationAccess(list, context, 'query');
+  if (!operationAccess) {
+    return null;
+  }
+
+  // TODO add item access control
```
Comment (dcousens) [inline packages/core/src/lib/core/queries/resolvers.ts:116]: As discussed during the review, we want to remove `findSingleton` and `updateSingleton`,  but notably can we add some tests for access control with Singletons

File: packages/core/src/lib/core/types-for-lists.ts [hunk 8]
```diff
@@ -84,13 +98,15 @@ function getIsEnabled(listsConfig: KeystoneConfig['lists']) {
...
+    const defaultIsFilterable = listConfig.kind === 'list' && listConfig?.defaultIsFilterable;
+    const defaultIsOrderable = listConfig.kind === 'list' && listConfig?.defaultIsOrderable;
+
     if (!omit) {
       // We explicity check for boolean/function values here to ensure the dev hasn't made a mistake
       // when defining these values. We avoid duck-typing here as this is security related
       // and we want to make it hard to write incorrect code.
       throwIfNotAFilter(defaultIsFilterable, listKey, 'defaultIsFilterable');
-      throwIfNotAFilter(defaultIsOrderable, listKey, 'defaultIsOrderable');
+      throwIfNotAFilter(defaultIsFilterable, listKey, 'defaultIsFilterable');
```
Comment (dcousens) [inline packages/core/src/lib/core/types-for-lists.ts:109]:
```suggestion
      throwIfNotAFilter(defaultIsOrderable, listKey, 'defaultIsOrderable');
```

File: packages/core/src/lib/core/types-for-lists.ts [hunk 9]
```diff
@@ -42,25 +44,37 @@ export type InitialisedField = Omit<NextFieldType, 'dbField' | 'access' | 'graph
...
-export type InitialisedList = {
+type CommonInitialisedList = {
   fields: Record<string, InitialisedField>;
   /** This will include the opposites to one-sided relationships */
   resolvedDbFields: Record<string, ResolvedDBField>;
-  pluralGraphQLName: string;
-  types: GraphQLTypesForList;
   access: ResolvedListAccessControl;
   hooks: ListHooks<BaseListTypeInfo>;
   adminUILabels: { label: string; singular: string; plural: string; path: string };
```
Comment (dcousens) [inline packages/core/src/lib/core/types-for-lists.ts:53]:
```suggestion
  adminUILabels: {
    label: string;
    singular: string; // TODO: remove for singletons
    plural: string; // TODO: remove for singletons
    path: string
  };
```

File: packages/core/src/lib/core/types-for-lists.ts [hunk 10]
```diff
@@ -42,25 +44,37 @@ export type InitialisedField = Omit<NextFieldType, 'dbField' | 'access' | 'graph
...
   cacheHint: ((args: CacheHintArgs) => CacheHint) | undefined;
```
Comment (dcousens) [inline packages/core/src/lib/core/types-for-lists.ts:54]:
```suggestion
  cacheHint: ((args: CacheHintArgs) => CacheHint) | undefined; // TODO: remove for singletons?
```

File: packages/core/src/lib/core/types-for-lists.ts [hunk 11]
```diff
@@ -42,25 +44,37 @@ export type InitialisedField = Omit<NextFieldType, 'dbField' | 'access' | 'graph
...
-  lists: Record<string, InitialisedList>;
+  lists: Record<string, InitialisedSchema>;
   dbMap: string | undefined;
+  pluralGraphQLName: string;
+  types: GraphQLTypesForList;
+  maxResults: number;
   graphql: {
     isEnabled: IsEnabled;
   };
 };
```
Comment (dcousens) [inline packages/core/src/lib/core/types-for-lists.ts:64]: Many of the options for a 'common list' aren't actually common to singletons, we should review this, maybe separately to the first merge

File: packages/core/src/lib/core/types-for-lists.ts [hunk 12]
```diff
@@ -108,19 +124,19 @@ function getIsEnabled(listsConfig: KeystoneConfig['lists']) {
         query: true,
         create: true,
         update: true,
-        delete: true,
-        filter: defaultIsFilterable ?? true,
-        orderBy: defaultIsOrderable ?? true,
+        delete: listConfig.kind === 'list',
+        filter: defaultIsFilterable ?? listConfig.kind === 'list',
+        orderBy: defaultIsOrderable ?? listConfig.kind === 'list',
```
Comment (dcousens) [inline packages/core/src/lib/core/types-for-lists.ts:129]: You are differentiating, but then not differentiating.  Either use `true` and ignore them, or set them to `false` for singletons (and then ignore them...)

File: packages/core/src/lib/core/types-for-lists.ts [hunk 13]
```diff
@@ -147,7 +163,7 @@ function getListsWithInitialisedFields(
               listKey,
               lists: listGraphqlTypes,
               provider,
-              getStorage: storage => configStorage?.[storage],
+              getStorage: (storage: string) => configStorage?.[storage],
```
Comment (dcousens) [inline packages/core/src/lib/core/types-for-lists.ts:166]:
```suggestion
              getStorage: storage => configStorage?.[storage],
```
Revert

File: packages/core/src/lib/core/types-for-lists.ts [hunk 14]
```diff
@@ -157,7 +173,7 @@ function getListsWithInitialisedFields(
             // when defining these values. We avoid duck-typing here as this is security related
             // and we want to make it hard to write incorrect code.
             throwIfNotAFilter(f.isFilterable, listKey, 'isFilterable');
-            throwIfNotAFilter(f.isOrderable, listKey, 'isOrderable');
+            throwIfNotAFilter(f.isFilterable, listKey, 'isOrderable');
```
Comment (dcousens) [inline packages/core/src/lib/core/types-for-lists.ts:176]:
```suggestion
            throwIfNotAFilter(f.isOrderable, listKey, 'isOrderable');
```

File: packages/core/src/lib/core/types-for-lists.ts [hunk 15]
```diff
@@ -451,8 +467,8 @@ export function initialiseLists(config: KeystoneConfig): Record<string, Initiali
    *
    * The object will be populated at the end of this function, and the reference will be maintained
    */
-  const listsRef: Record<string, InitialisedList> = {};
-
+  const listsRef: Record<string, InitialisedSchema> = {};
+  /** Block statements to contain variables only being used within them */
```
Comment (dcousens) [inline packages/core/src/lib/core/types-for-lists.ts:471]:
```suggestion
```

File: packages/core/src/lib/createSystem.ts [hunk 16]
```diff
@@ -102,6 +104,7 @@ export function createSystem(config: KeystoneConfig, isLiveReload?: boolean) {
           if (!isLiveReload) {
             await prismaClient.$connect();
             const context = createContext({ sudo: true });
+            await ensureSingletons(lists, prismaClient);
```
Comment (dcousens) [inline packages/core/src/lib/createSystem.ts:107]: This should not happen in production, we should automatically create a migration.

In development, we should _maybe_ throw if the constraints we rely on aren't being met, but, not in production.

File: packages/core/src/types/config/access-control.ts [hunk 17]
```diff
@@ -117,32 +122,54 @@ export type ListAccessControl<ListTypeInfo extends BaseListTypeInfo> = {
...
+    item: ListTypeInfo['item'];
+  };
+
+export type SingletonAccessControl<ListTypeInfo extends BaseSingletonTypeInfo> = {
+  // These functions should return `true` if access is allowed or `false` if access is denied.
+  operation?: {
+    query?: ListOperationAccessControl<'query', ListTypeInfo>;
+    update?: ListOperationAccessControl<'update', ListTypeInfo>;
+  };
+
```
Comment (dcousens) [inline packages/core/src/types/config/access-control.ts:140]:
```suggestion
  filter?: {
    query?: ListFilterAccessControl<'query', ListTypeInfo>;
    update?: ListFilterAccessControl<'update', ListTypeInfo>;
  };
```

File: packages/core/src/types/config/access-control.ts [hunk 18]
```diff
@@ -117,32 +122,54 @@ export type ListAccessControl<ListTypeInfo extends BaseListTypeInfo> = {
...
+  // These rules are applied to each item being operated on individually. They return `true` or `false`,
+  // and if false, an access denied error will be returned for the individual operation.
+  item?: {
+    read?: ReadItemAccessArgs<ListTypeInfo>;
```
Comment (dcousens) [inline packages/core/src/types/config/access-control.ts:144]:
```suggestion
```

File: packages/core/src/types/context.ts [hunk 19]
```diff
@@ -33,59 +33,75 @@ export type KeystoneContext<TypeInfo extends BaseKeystoneTypeInfo = BaseKeystone
...
-    ): Promise<Record<string, any> | null>;
-    deleteMany(
-      args: {
-        readonly where: readonly KeystoneListsTypeInfo[Key]['inputs']['uniqueWhere'][];
-      } & ResolveFields
-    ): Promise<Record<string, any>[]>;
-  };
+export type KeystoneSingletonAPI<ListTypeInfo extends BaseSingletonTypeInfo> = {
+  read(args: ResolveFields): Promise<Record<string, any>>;
+  update(
```
Comment (dcousens) [inline packages/core/src/types/context.ts:38]:
```suggestion
  findOne(args: ResolveFields): Promise<Record<string, any>>;
  updateOne(
```

File: packages/core/src/types/context.ts [hunk 20]
```diff
@@ -33,59 +33,75 @@ export type KeystoneContext<TypeInfo extends BaseKeystoneTypeInfo = BaseKeystone
...
-        readonly where: KeystoneListsTypeInfo[Key]['inputs']['uniqueWhere'];
-      } & ResolveFields
-    ): Promise<Record<string, any> | null>;
-    deleteMany(
-      args: {
-        readonly where: readonly KeystoneListsTypeInfo[Key]['inputs']['uniqueWhere'][];
-      } & ResolveFields
-    ): Promise<Record<string, any>[]>;
-  };
+export type KeystoneSingletonAPI<ListTypeInfo extends BaseSingletonTypeInfo> = {
```
Comment (dcousens) [inline packages/core/src/types/context.ts:36]:
```suggestion
export type KeystoneSingletonAPI<ListTypeInfo extends BaseSingletonTypeInfo> = {
  kind: 'singleton',
```

File: packages/core/src/types/context.ts [hunk 21]
```diff
@@ -33,59 +33,75 @@ export type KeystoneContext<TypeInfo extends BaseKeystoneTypeInfo = BaseKeystone
...
+    } & ResolveFields
+  ): Promise<Record<string, any> | null>;
+  deleteMany(
+    args: {
+      readonly where: readonly ListTypeInfo['inputs']['uniqueWhere'][];
+    } & ResolveFields
+  ): Promise<Record<string, any>[]>;
+};
+
+type KeystoneListAPI<ListTypeInfo extends BaseSchemaTypeInfo> =
```
Comment (dcousens) [inline packages/core/src/types/context.ts:96]:
```suggestion
type KeystoneQueryAPI<ListTypeInfo extends BaseSchemaTypeInfo> =
```

File: packages/core/src/types/context.ts [hunk 22]
```diff
@@ -33,59 +33,75 @@ export type KeystoneContext<TypeInfo extends BaseKeystoneTypeInfo = BaseKeystone
...
+  ListTypeInfo extends BaseListTypeInfo
+    ? KeystoneStandardListsAPI<ListTypeInfo>
```
Comment (dcousens) [inline packages/core/src/types/context.ts:98]:
```suggestion
    ? KeystoneListsAPI<ListTypeInfo>
```

File: packages/core/src/types/context.ts [hunk 23]
```diff
@@ -95,45 +111,59 @@ type ResolveFields = {
...
+  }): Promise<ListTypeInfo['item'][]>;
+};
+
+type KeystoneSingletonDbAPI<ListTypeInfo extends BaseSingletonTypeInfo> = {
+  kind: 'singleton';
+  read(): Promise<ListTypeInfo['item']>;
+  update(args: { readonly data: ListTypeInfo['inputs']['update'] }): Promise<ListTypeInfo['item']>;
+};
+
+export type KeystoneIndividualDbAPI<ListTypeInfo extends BaseSchemaTypeInfo> =
```
Comment (dcousens) [inline packages/core/src/types/context.ts:158]: This name sucks

File: packages/core/src/types/next-fields.ts [hunk 24]
```diff
@@ -393,6 +393,25 @@ export type FieldTypeWithoutDBField<
 
 type AnyInputObj = graphql.InputObjectType<Record<string, graphql.Arg<graphql.InputType, any>>>;
 
+export type GraphqlTypesForSingleton = {
+  update: AnyInputObj;
+  create: AnyInputObj;
```
Comment (dcousens) [inline packages/core/src/types/next-fields.ts:398]:
```suggestion
```

File: packages/core/src/types/next-fields.ts [hunk 25]
```diff
@@ -393,6 +393,25 @@ export type FieldTypeWithoutDBField<
...
+        connect: graphql.Arg<typeof graphql.Boolean>;
+      }>;
+      update?: graphql.InputObjectType<{
+        create?: graphql.Arg<GraphqlTypesForSingleton['create']>;
+        connect: graphql.Arg<typeof graphql.Boolean>;
+        disconnect: graphql.Arg<typeof graphql.Boolean>;
+      }>;
+    };
+  };
+};
```
Comment (dcousens) [inline packages/core/src/types/next-fields.ts:413]:
```suggestion
export type GraphqlTypesForSingleton = {
  update: AnyInputObj;
  output: graphql.ObjectType<BaseItem>;
};
```

File: packages/core/src/types/next-fields.ts [hunk 26]
```diff
@@ -393,6 +393,25 @@ export type FieldTypeWithoutDBField<
...
+
 export type GraphQLTypesForList = {
```
Comment (dcousens) [inline packages/core/src/types/next-fields.ts:415]: kind: 'list'

Comment (dcousens) [inline packages/core/src/types/next-fields.ts:415]:
```suggestion
export type GraphQLTypesForList = {
  kind: 'list'
```

File: packages/core/src/types/type-info.ts [hunk 27]
```diff
@@ -3,7 +3,20 @@ import { BaseItem } from './next-fields';
...
 type GraphQLInput = Record<string, any>;
 
+export type BaseSingletonTypeInfo = {
+  kind: 'singleton';
+  key: string;
+  fields: string;
+  item: BaseItem;
+  inputs: {
+    update: GraphQLInput;
+    create: never;
```
Comment (dcousens) [inline packages/core/src/types/type-info.ts:13]:
```suggestion
  inputs: {
    create: never; // this is complicated, but it shouldn't exist
    update: GraphQLInput;
```

File: packages/auth/src/gql/getPasswordResetSchema.ts [hunk 28]
```diff
@@ -45,9 +50,8 @@ export function getPasswordResetSchema<I extends string, S extends string>({
         type: graphql.nonNull(graphql.Boolean),
         args: { [identityField]: graphql.arg({ type: graphql.nonNull(graphql.String) }) },
         async resolve(rootVal, { [identityField]: identity }, context) {
-          const dbItemAPI = context.sudo().db[listKey];
+          const dbItemAPI = (context as ContextWithOnlyStandardLists).sudo().db[listKey];
```
Comment (dcousens) [inline packages/auth/src/gql/getPasswordResetSchema.ts:53]:
```suggestion
          const dbItemAPI = getContextDb(context.sudo(), listKey);
```
And remove the `as ContextWithOnlyStandardLists` types

File: packages/core/src/admin-ui/system/getAdminMetaSchema.ts [hunk 29]
```diff
@@ -251,26 +261,68 @@ export function getAdminMetaSchema({
...
+    }),
+  });
+
+  const names = {
+    singleton: 'KeystoneAdminUISingletonMeta',
+    list: 'KeystoneAdminUIListMeta',
+  };
+
+  const KeystoneAdminUISchemaMeta = graphql.interface<SchemaMetaRootVal>()({
+    name: 'KeystoneAdminUISchemaMeta',
```
Comment (dcousens) [inline packages/core/src/admin-ui/system/getAdminMetaSchema.ts:296]:
```suggestion
    name: 'KeystoneAdminUISchemaTypeMeta',
```

File: packages/core/src/admin-ui/system/getAdminMetaSchema.ts [hunk 30]
```diff
@@ -281,45 +333,22 @@ export function getAdminMetaSchema({
         resolve(rootVal, args, context) {
           if ('isAdminUIBuildProcess' in context) {
             throw new Error(
-              'KeystoneAdminUIListMeta.hideDelete cannot be resolved during the build process'
+              'KeystoneAdminUISchemaMeta.hideDelete cannot be resolved during the build process'
```
Comment (dcousens) [inline packages/core/src/admin-ui/system/getAdminMetaSchema.ts:336]:
```suggestion
              'KeystoneAdminUISchemaTypeMeta.hideDelete cannot be resolved during the build process'
```

File: packages/core/src/fields/types/relationship/views/index.tsx [hunk 31]
```diff
@@ -623,10 +625,12 @@ export const controller = (
   };
 };
 
-function useRelationshipFilterValues({ value, list }: { value: string; list: ListMeta }) {
+function useRelationshipFilterValues({ value, list }: { value: string; list: SchemaMeta }) {
   const foreignIds = getForeignIds(value);
   const where = { id: { in: foreignIds } };
-
+  if (list.kind === 'singleton') {
+    throw new Error('Singleton bad');
```
Comment (dcousens) [inline packages/core/src/fields/types/relationship/views/index.tsx:632]: Bad :name_badge:

File: packages/core/src/fields/types/relationship/views/index.tsx [hunk 32]
```diff
@@ -623,10 +625,12 @@ export const controller = (
   };
 };
 
-function useRelationshipFilterValues({ value, list }: { value: string; list: ListMeta }) {
+function useRelationshipFilterValues({ value, list }: { value: string; list: SchemaMeta }) {
   const foreignIds = getForeignIds(value);
   const where = { id: { in: foreignIds } };
-
+  if (list.kind === 'singleton') {
```
Comment (dcousens) [inline packages/core/src/fields/types/relationship/views/index.tsx:631]:
```suggestion
  if (list.kind !== 'list') {
```

File: packages/core/src/fields/types/relationship/views/index.tsx [hunk 33]
```diff
@@ -623,10 +625,12 @@ export const controller = (
...
+    throw new Error('Singleton bad');
+  }
```
Comment (dcousens) [inline packages/core/src/fields/types/relationship/views/index.tsx:633]:
```suggestion
  if (schema.kind !== 'list') {
    throw new Error('An actual error');
  }
```

File: packages/core/src/fields/types/timestamp/tests/test-fixtures.ts [hunk 34]
```diff
@@ -33,77 +32,3 @@ export const storedValues = () => [
   { name: 'person6', lastOnline: null },
   { name: 'person7', lastOnline: null },
 ];
-
-export const filterTests = (withKeystone: (args: any) => any) => {
```
Comment (dcousens) [inline packages/core/src/fields/types/timestamp/tests/test-fixtures.ts:37]: These tests are not running at present, and we aren't going to change that in this pull request.

File: packages/core/src/___internal-do-not-use-will-break-in-patch/admin-ui/pages/SingletonPage/index.tsx [hunk 35]
```diff
@@ -0,0 +1,372 @@
...
+        }}
+        isOpen={isConfirmModalOpen}
+        title="Are you sure you want to reset changes?"
+        tone="negative"
+      >
+        {null}
+      </AlertDialog>
+    </Fragment>
+  );
+}
```
Comment (mitchellhamilton) [inline packages/core/src/___internal-do-not-use-will-break-in-patch/admin-ui/pages/SingletonPage/index.tsx:372]: Let's share this with the list version

File: packages/core/src/lib/config/initConfig.ts [hunk 36]
```diff
@@ -7,7 +7,7 @@ import { applyIdFieldDefaults } from './applyIdFieldDefaults';
   needs to be run on Keystone Config before it can be used.
 */
 
-export function initConfig(config: KeystoneConfig) {
+export function initConfig(config: KeystoneConfig<any>) {
```
Comment (dcousens) [inline packages/core/src/lib/config/initConfig.ts:10]:
```suggestion
export function initConfig(config: KeystoneConfig) {
```

File: packages/core/src/___internal-do-not-use-will-break-in-patch/admin-ui/pages/SingletonPage/index.tsx (hunk 35 above)
Comment (mitchellhamilton) [inline packages/core/src/___internal-do-not-use-will-break-in-patch/admin-ui/pages/SingletonPage/index.tsx:372]: Let's share this with the list version

File: packages/core/src/lib/context/itemAPI.ts [hunk 37]
```diff
@@ -5,20 +5,16 @@ import {
...
-const objectEntriesButUsingKeyof: <T extends Record<string, any>>(
-  obj: T
-) => [keyof T, T[keyof T]][] = Object.entries as any;
-
 export function getDbAPIFactory(
   gqlNames: GqlNames,
-  schema: GraphQLSchema
-): (context: KeystoneContext) => KeystoneDbAPI<Record<string, BaseListTypeInfo>>[string] {
+  schema: GraphQLSchema,
+  kind: 'list' | 'singleton'
```
Comment (dcousens) [inline packages/core/src/lib/context/itemAPI.ts:16]:
```suggestion
  kind: BaseSchemaTypeInfo['kind']
```

File: tests/api-tests/utils.ts [hunk 38]
```diff
@@ -1,22 +1,77 @@
...
+  ? TypeInfo
+  : never;
+
 // This function injects the db configuration that we use for testing in CI.
 // This functionality is a keystone repo specific way of doing things, so we don't
 // export it from `@keystone-6/core/testing`.
-export const apiTestConfig = (
-  config: Omit<KeystoneConfig, 'db'> & {
+export const apiTestConfig = <Lists extends ListsConfig>(
+  config: Omit<KeystoneConfig, 'db' | 'lists'> & {
```
Comment (dcousens) [inline tests/api-tests/utils.ts:45]:
```suggestion
  config: Omit<KeystoneConfig, 'db' | 'schema'> & {
```

File: tests/api-tests/utils.ts [hunk 39]
```diff
@@ -98,7 +153,7 @@ export const expectExtensionError = (
   const unpackedErrors = unpackErrors(errors);
   expect(unpackedErrors).toEqual(
     args.map(({ path, messages, debug }) => {
-      const message = `An error occured while running "${extensionName}".\n${j(messages)}`;
+      const message: string = `An error occured while running "${extensionName}".\n${j(messages)}`;
```
Comment (dcousens) [inline tests/api-tests/utils.ts:156]:
```suggestion
      const message = `An error occured while running "${extensionName}".\n${j(messages)}`;
```

**Comment by Noviny** [comment]
Replaced by #7863

**Pull request closed without merging**